### PR TITLE
Implement Add-PoshGitToProfile command

### PR DESCRIPTION
**Pull request by rkeithhill**
Offer to update their current/all host profile if posh-git not detected in their profile scripts.  Note: we do not offer to update the "all users" profile scripts because that requires elevation and probably not worth it.

Add Pester tests for the new Utils.ps1 functions.

### CODE REVIEW

File: posh-git.psm1 [hunk 1]
```diff
@@ -92,6 +92,57 @@ if (!$currentPromptDef -or ($currentPromptDef -eq $defaultPromptDef)) {
     Set-Item Function:\prompt -Value $poshGitPromptScriptBlock
 }
 
+# If running interactive, check if user want's `import-module posh-git` added to their profile
+if (!$NoProfileCheck -and ($MyInvocation.ScriptName.Length -eq 0)) {
```
Comment (dahlbyk) [inline posh-git.psm1]: Is the `Length` check more idiomatic than `!$MyInvocation.ScriptName`?

Also, s/want's/wants above. :trollface:

File: test/Utils.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+                $scriptRoot = Split-Path $profilePath -Parent
+                Add-ImportModuleToProfile $profilePath $scriptRoot
+                Test-Path -LiteralPath $profilePath | Should Be $true
+                Get-FileEncoding $profilePath | Should Be 'ascii'
+                $content = Get-Content $profilePath
+                $content.Count | Should Be 1
+                @($content)[0] | Should BeExactly "Import-Module '$scriptRoot\posh-git.psd1'"
+            }
+            finally {
+                Remove-Item $profilePath -ErrorAction SilentlyContinue
```
Comment (dahlbyk) [inline test/Utils.Tests.ps1]:
```powershell
BeforeEach {
    $profilePath = [System.IO.Path]::GetTempFileName()
}
AfterEach {
    Remove-Item $profilePath -ErrorAction SilentlyContinue
}
```
❓

File: Utils.ps1 [hunk 3]
```diff
@@ -26,6 +27,98 @@ function Invoke-Utf8ConsoleCommand ([ScriptBlock]$cmd) {
...
+    if (Test-Path -LiteralPath $profilePath) {
+        $profileContent = @(Get-Content -LiteralPath $profilePath)
+        $profileContent += [string]::Empty # Empty line between previous profile script and import statement
+        $encoding = Get-FileEncoding $profilePath
+    }
+    else {
+        # Doesn't exist, so create it
+        New-Item -Path $profilePath -ItemType File
+        $profileContent = @()
+        $encoding = 'ascii'
```
Comment (rkeithhill) [inline Utils.ps1]: Check.

File: Utils.ps1 [hunk 4]
```diff
@@ -26,6 +27,98 @@ function Invoke-Utf8ConsoleCommand ([ScriptBlock]$cmd) {
...
+    }
+
+    # Check if the location of this module file is in the PSModulePath
+    if (Test-InModulePath $ScriptRoot) {
+        $profileContent += "Import-Module posh-git"
```
Comment (rkeithhill) [inline Utils.ps1]: I would avoid hard-coding the module path as a "bad practice" for PowerShell modules. I can switch to using Get-Module -ListAvailable.

File: install.ps1 [hunk 5]
```diff
@@ -1,5 +1,8 @@
 param([switch]$WhatIf = $false)
 
+# Dot source for Get-FileEncoding
+. $PSScriptRoot\Utils.ps1
```
Comment (rkeithhill) [inline install.ps1]: Thought I caught that.  We can just use Split-Path $MyInvocation.MyCommand.Path -Parent instead.

File: test/Utils.Tests.ps1 (hunk 2 above)
Comment (rkeithhill) [inline test/Utils.Tests.ps1]: Heh, I started with that.  Then wrote a few tests that didn't need $profilePath and removed Before/AfterEach.  Then I moved those tests to a different Context.  Fixed.

File: Utils.ps1 (hunk 4 above)
Comment (rkeithhill) [inline Utils.ps1]: Doh, that makes the Pester test a little harder.  Oh well, hopefully not too much harder.  :-)

Comment (rkeithhill) [inline Utils.ps1]: OK this is making the Pester tests a lot harder.  For some reason, Mock Get-Module is not working.  Will look at it more tomorrow.

File: install.ps1 [hunk 6]
```diff
@@ -1,5 +1,9 @@
 param([switch]$WhatIf = $false)
 
+# Dot source for Get-FileEncoding
+$scriptRoot = Split-Path $MyInvocation.MyCommand.Path -Parent
```
Comment (dahlbyk) [inline install.ps1]: We have `$installDir = Split-Path $MyInvocation.MyCommand.Path -Parent` on L22. Rename this one and delete that line?

**Comment by dahlbyk** [comment]
Minor nitpick, otherwise revisions look good. Just need to fix up how we detect if we should `Import-Module posh-git`!

**Comment by dahlbyk** [comment]
Is this ready for a final review? (Go ahead and remove WIP from title if so.)

**Comment by rkeithhill** [comment]
Yes. Ready for final review.

**Comment by rkeithhill** [comment]
Hmm, hold off on this a bit.  Joel says ScriptName is empty in implicit loading scenarios and he's right.  Let's see if we can use Get-PSCallstack instead.

**Comment by rkeithhill** [comment]
I'm becoming quite nervous about this PR.  Using `Get-PSCallstack` has its issues as well.  In general I thought a stack count of 2 would be reliably indicative of importing from the prompt e.g. here is a manual import call stack:
```
PS C:\Users\Keith> ipmo posh-git

Command          : posh-git.psm1
Location         : posh-git.psm1: line 95
Arguments        : {}
ScriptName       : C:\Users\Keith\Documents\WindowsPowerShell\Modules\posh-git\0.7.0\posh-git.psm1
ScriptLineNumber : 95
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : Get-PSCallstack | fl * | Out-String
FunctionName     : <ScriptBlock>

Command          : <ScriptBlock>
Location         : <No file>
Arguments        : {}
ScriptName       :
ScriptLineNumber : 1
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : ipmo posh-git
FunctionName     : <ScriptBlock>
```
And when the PowerShell console profile runs, the stack depth is three:
```
Command          : posh-git.psm1
Location         : posh-git.psm1: line 95
Arguments        : {}
ScriptName       : C:\Users\Keith\Documents\WindowsPowerShell\Modules\posh-git\0.7.0\posh-git.psm1
ScriptLineNumber : 95
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : Get-PSCallstack | fl * | Out-String
FunctionName     : <ScriptBlock>

Command          : profile.ps1
Location         : profile.ps1: line 9
Arguments        : {}
ScriptName       : C:\Users\Keith\Documents\WindowsPowerShell\profile.ps1
ScriptLineNumber : 9
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : Import-Module posh-git
FunctionName     : <ScriptBlock>

Command          : <ScriptBlock>
Location         : <No file>
Arguments        : {}
ScriptName       :
ScriptLineNumber : 1
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : . 'C:\Users\Keith\Documents\WindowsPowerShell\profile.ps1'
FunctionName     : <ScriptBlock>
```
But then I started to notice that the PowerShell in VSCode on my machine keep stalling during init.  And of course it was stalling because of the query to modify the import.  Sigh...  There is something about the PowerShell extension custom host that results in a stack depth one less than expected when processing profiles e.g.:
```
Command          : posh-git.psm1
Location         : posh-git.psm1: line 96
Arguments        : {}
ScriptName       : C:\\Users\\Keith\\Documents\\WindowsPowerShell\\Modules\\posh-git\\0.7.0\\posh-git.psm1
ScriptLineNumber : 96
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : Get-PSCallstack | Fl * | Out-String
FunctionName     : <ScriptBlock>

Command          : profile.ps1
Location         : profile.ps1: line 10
Arguments        : {}
ScriptName       : C:\\Users\\Keith\\Documents\\WindowsPowerShell\\profile.ps1
ScriptLineNumber : 10
InvocationInfo   : System.Management.Automation.InvocationInfo
Position         : Get-GitStatus
FunctionName     : <ScriptBlock>
```
Another issue is that if posh-git hasn't been imported yet and you execute one of its commands, it will be auto-imported.  That will cause the Modify Profile prompt to appear. That could be confusing.

How about we rejigger this work into a command like `Add-PoshGitToProfile`?  It will result in a pretty simple install process e.g.:
```powershell
Install-Module posh-git -Scope User
Add-PoshGitToProfile
```
Otherwise I'm worried there will be some unanticipated scenario (other custom hosts, remoting, etc) that will result in posh-git stalling the script waiting for input.

**Comment by rkeithhill** [comment]
Looking that the Docker instructions for [getting going with the posh-docker PowerShell module](https://docs.docker.com/docker-for-windows) see the section `Set up tab completion in PowerShell`.  Their profile mod approach is simply:
```powershell
Install-Module -Scope CurrentUser posh-docker -Force
Add-Content $PROFILE "`nImport-Module posh-docker"
```
I like the simplicity.  They could have also mentioned that if you want posh-docker loaded in all hosts, then use:
```powershell
Install-Module -Scope CurrentUser posh-docker -Force
Add-Content $PROFILE.CurrentUserAllHosts "`nImport-Module posh-docker"
```
And Add-Content seems to preserve the original file's encoding.

**Comment by dahlbyk** [comment]
If you're nervous, I'm nervous. Let's skip the magic. I appreciate the simplicity of Docker's approach, but wouldn't mind being a bit more helpful. How about:

`Add-PoshGitToProfile [-AllHosts] [-Force]`

- `-AllHosts` would switch from `CurrentUserCurrentHost` to `CurrentUserAllHosts`.
- `-Force` would skip our check if posh-git is already being imported by a profile.
- We would keep our check to see if we can import by name or need to import the full path.
- The Chocolatey install would replace its complexity with an import and `Add-PoshGitToProfile`

> And Add-Content seems to preserve the original file's encoding.

And specifying `-Encoding UTF8` seems to provide the encoding to use if the file doesn't exist. :100:

We'll still need to handle encoding (we don't today) in Chocolatey install/uninstall, but `Add-Content` seems sufficient for `Add-PoshGitToProfile`.

**Comment by rkeithhill** [comment]
> I appreciate the simplicity of Docker's approach, but wouldn't mind being a bit more helpful.

Agreed.  Everything else sounds good as well.  I'm starting on it.

**Comment by rkeithhill** [comment]
I'm ready to merge this.

**Pull request merged**